cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/program/MessagingProgramStateWriter.java

```diff
@@ -78,106 +92,102 @@ public void start(ProgramRunId programRunId, ProgramOptions programOptions, @Nul
+    scheduleHeartBeatThread(properties.build());
+  }
+
+  /**
+   * If executor service isn't initialized or if its shutdown
+   * create a new exector service and schedule a heartbeat thread
+   * @param properties
+   */
+  private void scheduleHeartBeatThread(Map<String, String> properties) {
+    if (scheduledExecutorService == null || scheduledExecutorService.isShutdown()) {
+      scheduledExecutorService =
+        Executors.newSingleThreadScheduledExecutor(Threads.createDaemonThreadFactory("program-heart-beat"));
+      ProgramRunHeartbeat programRunHeartbeat = new ProgramRunHeartbeat(programStatePublisher, properties);
+      scheduledExecutorService.scheduleAtFixedRate(programRunHeartbeat, heartBeatIntervalSeconds,
+                                                   heartBeatIntervalSeconds, TimeUnit.SECONDS);
+    }
   }
 
   @Override
-  public void completed(ProgramRunId programRunId) {
-    stop(programRunId, ProgramRunStatus.COMPLETED, null);
+  public void completed(ProgramRunId programRunId, ProgramOptions programOptions) {
+    stop(programRunId, ProgramRunStatus.COMPLETED, null, programOptions);
   }
 
   @Override
-  public void killed(ProgramRunId programRunId) {
-    stop(programRunId, ProgramRunStatus.KILLED, null);
+  public void killed(ProgramRunId programRunId, ProgramOptions programOptions) {
+    stop(programRunId, ProgramRunStatus.KILLED, null, programOptions);
   }
 
   @Override
-  public void error(ProgramRunId programRunId, Throwable failureCause) {
-    stop(programRunId, ProgramRunStatus.FAILED, failureCause);
+  public void error(ProgramRunId programRunId, Throwable failureCause, ProgramOptions programOptions) {
+    stop(programRunId, ProgramRunStatus.FAILED, failureCause, programOptions);
   }
 
   @Override
-  public void suspend(ProgramRunId programRunId) {
-    publish(
-      ImmutableMap.<String, String>builder()
-        .put(ProgramOptionConstants.PROGRAM_RUN_ID, GSON.toJson(programRunId))
-        .put(ProgramOptionConstants.SUSPEND_TIME, String.valueOf(System.currentTimeMillis()))
-        .put(ProgramOptionConstants.PROGRAM_STATUS, ProgramRunStatus.SUSPENDED.name())
+  public void suspend(ProgramRunId programRunId, ProgramOptions programOptions) {
+    programStatePublisher.publish(Notification.Type.PROGRAM_STATUS,
+                                  ImmutableMap.<String, String>builder()
+                                    .put(ProgramOptionConstants.PROGRAM_RUN_ID, GSON.toJson(programRunId))
+                                    .put(ProgramOptionConstants.SUSPEND_TIME,
+                                         String.valueOf(System.currentTimeMillis()))
+                                    .put(ProgramOptionConstants.PROGRAM_STATUS, ProgramRunStatus.SUSPENDED.name())
+                                    .put(ProgramOptionConstants.PROGRAM_OPTIONS, GSON.toJson(programOptions)).build()
     );
+    if (scheduledExecutorService != null) {
```

Comment:
you should shutdown before you publish, otherwise you can still sent other messages after publishing